contracts/zero-ex/contracts/src/features/multiplex/MultiplexUniswapV2.sol

```diff
@@ -0,0 +1,256 @@
+    FixinTokenSpender
+{
+    using LibSafeMathV06 for uint256;
+
+    // address of the UniswapV2Factory contract.
+    address private immutable UNISWAP_FACTORY;
+    // address of the (Sushiswap) UniswapV2Factory contract.
+    address private immutable SUSHISWAP_FACTORY;
+    // Init code hash of the UniswapV2Pair contract.
+    bytes32 private immutable UNISWAP_PAIR_INIT_CODE_HASH;
+    // Init code hash of the (Sushiswap) UniswapV2Pair contract.
+    bytes32 private immutable SUSHISWAP_PAIR_INIT_CODE_HASH;
+
+    constructor(
+        address uniswapFactory,
+        address sushiswapFactory,
+        bytes32 uniswapPairInitCodeHash,
+        bytes32 sushiswapPairInitCodeHash
+    )
+        internal
+    {
+        UNISWAP_FACTORY = uniswapFactory;
+        SUSHISWAP_FACTORY = sushiswapFactory;
+        UNISWAP_PAIR_INIT_CODE_HASH = uniswapPairInitCodeHash;
+        SUSHISWAP_PAIR_INIT_CODE_HASH = sushiswapPairInitCodeHash;
+    }
+
+    function _batchSellUniswapV2(
+        IMultiplexFeature.BatchSellState memory state,
+        IMultiplexFeature.BatchSellParams memory params,
+        bytes memory wrappedCallData,
+        uint256 sellAmount
+    )
+        internal
+    {
+        (address[] memory tokens, bool isSushi) = abi.decode(
+            wrappedCallData,
+            (address[], bool)
+        );
+        // Validate tokens
+        require(
+            tokens.length >= 2 &&
+            tokens[0] == address(params.inputToken) &&
+            tokens[tokens.length - 1] == address(params.outputToken),
+            "MultiplexUniswapV2::_batchSellUniswapV2/INVALID_TOKENS"
+        );
+        // Compute the address of the first Uniswap pair
+        // contract that will execute a swap.
+        address firstPairAddress = _computeUniswapPairAddress(
+            tokens[0],
+            tokens[1],
+            isSushi
+        );
+        // `_sellToUniswapV2` assumes the input tokens have been
+        // transferred into the pair contract before it is called,
+        // so we transfer the tokens in now (either from `msg.sender`
+        // or using the Exchange Proxy's balance).
+        if (params.useSelfBalance) {
+            _transferERC20Tokens(
+                IERC20TokenV06(tokens[0]),
+                firstPairAddress,
+                sellAmount
+            );
+        } else {
+            _transferERC20TokensFrom(
+                IERC20TokenV06(tokens[0]),
+                msg.sender,
+                firstPairAddress,
+                sellAmount
+            );
+        }
+        // Execute the Uniswap/Sushiswap trade.
+        uint256 boughtAmount = _sellToUniswapV2(
+            tokens,
+            sellAmount,
+            isSushi,
+            firstPairAddress,
+            params.recipient
+        );
+        // Increment the sold and bought amounts.
+        state.soldAmount = state.soldAmount.safeAdd(sellAmount);
+        state.boughtAmount = state.boughtAmount.safeAdd(boughtAmount);
+    }
```

Comment:
oh yeah it's bc we transfer the input tokens into the pair contract before calling `swap`, so even if we swallow the revert here those tokens would still be gone

Reply:
Should we just not swallow reverts in the other places? Or maybe even wrap this all in a `delegatecall()` to revert the transfer?

Reply:
We definitely want to swallow reverts for RFQ. 
No real reason to swallow reverts for transformERC20, since we should only be using it for fallbacks, so we're very brittle to failures anyway. 
Uniswap V2 and V3 should only revert if there's not enough liquidity in a pool I think, so we could go either way on those.
All bets are off for LiquidityProvider, so maybe we should wrap it in a `delegatecall`? 

Are you thinking that we should either swallow reverts everywhere or not swallow reverts anywhere (except RFQ)? I don't have a strong opinion, I think we'd generally be pretty brittle to failures either way

Reply:
Yeah, I think it makes it easier to understand how these functions are intended to be used and makes routing bugs easier to suss out if the behavior is consistent.